pgxn/neon/neon.c

```diff
@@ -267,6 +268,148 @@ LogicalSlotsMonitorMain(Datum main_arg)
 	}
 }
 
+/*
+ * This is taken from procarry.c
+ * TODO: should we better move it to some header file?
+ */
+#define PROCARRAY_MAXPROCS	(MaxBackends + max_prepared_xacts)
+#define TOTAL_MAX_CACHED_SUBXIDS \
+	((PGPROC_MAX_CACHED_SUBXIDS + 1) * PROCARRAY_MAXPROCS)
+
+/*
+ * We do not want to wait running-xacts record from primary which is
+ * generated each 15 seconds (if there is some activity).  So we try
+ * to restore information about running transactions from CLOG.
+ */
+/*
+ * NEON: we do not want to wait running-xacts record from primary which is generated each 15 seconds (if there is some activity).
+ * So we try to restore information about running transactions from CLOG.
+ *
+ * This functions allocates and fill array at *xids and store number of restored xacts in nxids.
+ * But not more than TOTAL_MAX_CACHED_SUBXIDS / 2 is restored.
+ * In case of overflow, false is returned.
+ */
+static bool
+RestoreRunningXactsFromClog(CheckPoint *checkpoint, TransactionId **xids, int *nxids)
+{
+	int			xcnt;
+	int			max_xcnt;
+	TransactionId xid;
+	XLogRecPtr	xidlsn;
+	XidStatus	xidstatus;
+	bool		overflow = false;
+
+	TransactionId from = TransactionIdPrecedes(checkpoint->oldestXid, checkpoint->oldestActiveXid)
+		? checkpoint->oldestActiveXid : checkpoint->oldestXid;
```

Comment:
oldestXid should _always_ precede oldestActiveXid.  Unless oldestActiveXid is invalid, but let's handle that case explicitly, for the sake of readability.

Using oldestXid is not great in any case. It's correct, which is good. But it can be very old indeed, `vacuum_freeze_min_age` defaults to 50 million. That might still work out in many cases in practice, because there should still be very few, if any, XIDs that don't have the CLOG bits set. Although I wonder how long it takes to scan 50 million XIDs in the CLOG?

Another problem is that if there _are_ a lot of XIDs without CLOG bits set, so that this overflows, then the standby cannot start. We have that problem in general, of course, but if you start from oldestXid rather than a more up-to-date oldestActiveXid, it gets worse. I added a test case to demonstrate that problem to this branch. It starts 100000 subtransaction in the primary, and then kills the primary without writing commit/abort records for them. Then the primary is restarted, and then without doing much else, shut down again, this time cleanly. Then you start a replica. The replica starts with oldestActiveXid==0, because of the clean shut down. The CLOG-scanning therefore starts from oldestXid, which means it will find all those 100000 crashed sub-XIDs, and cannot start up. This problem would persist even if you start and stop the primary again, until the whole closer has been vacuum frozen, advancing oldestXid.

Reply:
Let's find a better mechanism for getting a valid starting point. For example:

In the pageserver, we could not reset oldestActiveXid to 0, when we ingest a checkpoint shutdown record. We could reset it to nextXid instead, for example. (But check that that does the right thing in the presence of prepared transactions.)

Reply:
Done